v 1.6.1 running only the preprocessor

on OSX I can do this:
```
$ eosio-cpp x.cpp
```
but I can no longer do this (I used to be able to, in v1.5.0):
```
$ eosio-cpp x.cpp -E -dD
```
I get the following errors:

> Error while trying to load a compilation database:
> Could not auto-detect compilation database for file "x.cpp"
> No compilation database found in /Users/ekkis/dev/eos or any parent directory
> fixed-compilation-database: Error while opening fixed database: No such file or directory
> json-compilation-database: Error while opening JSON database: No such file or directory
> Running without flags.
> /Users/ekkis/dev/eos/x.cpp:1:10: fatal error: 'eosio/eosio.hpp' file not found
> #include <eosio/eosio.hpp>
>          ^~~~~~~~~~~~~~~~~
> Warning, empty ricardian clause file
> Warning, empty ricardian clause file
> Warning, action <t> does not have a ricardian contract
> 1 error generated.
> Error while processing /Users/ekkis/dev/eos/x.cpp.
> abigen error

these are useful flags as they allow me to check for preprocessor-generated code, but now it's broken

I also meet the same problem,on `OS X`, I can compile with:
```bash
eosio-cpp hello.cpp -o hello.wasm
```
But I want to view macro extened by `preprocessor` with option `-E`:
```bash
eosio-cpp -E hello.cpp
``` 
error reported with:
```bash
/Users/fospring/work/cpp/src/github.com/EOSIO/test_contract/hello/hello.cpp:1:10: fatal error:
      'eosiolib/eosio.hpp' file not found
#include <eosiolib/eosio.hpp>
         ^~~~~~~~~~~~~~~~~~~~
Warning, empty ricardian clause file
Warning, empty ricardian clause file
Warning, action <hi> does not have a ricardian contract
1 error generated.
Error while processing /Users/fospring/work/cpp/src/github.com/EOSIO/test_contract/hello/hello.cpp.
abigen error
```

--- a/tools/cc/eosio-cc.cpp.in
+++ b/tools/cc/eosio-cc.cpp.in
@@ -58,7 +58,7 @@ int main(int argc, const char **argv) {
          output = opts.output_fn.empty() ? "a.out" : opts.output_fn;
       }
 
-      new_opts.insert(new_opts.begin(), "-o "+output);
+      new_opts.insert(new_opts.begin(), "-o"+output);
       outputs.push_back(output);
       if (!eosio::cdt::environment::exec_subprogram("clang-7", new_opts)) {
          llvm::sys::fs::remove(tmp_file);
@@ -71,7 +71,7 @@ int main(int argc, const char **argv) {
    if (opts.link) {
       std::vector<std::string> new_opts = opts.ld_options;
       for (auto input : outputs) {
-         new_opts.insert(new_opts.begin(), std::string(" ")+input+" ");
+         new_opts.insert(new_opts.begin(), input);
       }
       if (!eosio::cdt::environment::exec_subprogram("eosio-ld", new_opts)) {
          for (auto input : outputs) {

--- a/tools/cc/eosio-cpp.cpp.in
+++ b/tools/cc/eosio-cpp.cpp.in
@@ -201,7 +201,7 @@ int main(int argc, const char **argv) {
             output = opts.output_fn.empty() ? "a.out" : opts.output_fn;
          }
 
-         new_opts.insert(new_opts.begin(), "-o "+output);
+         new_opts.insert(new_opts.begin(), {"-o", output});
          outputs.push_back(output);
 
          if (llvm::sys::path::extension(input).equals(".c"))
@@ -221,7 +221,7 @@ int main(int argc, const char **argv) {
    if (opts.link) {
       std::vector<std::string> new_opts = opts.ld_options;
       for (auto input : outputs) {
-         new_opts.insert(new_opts.begin(), std::string(" ")+input+" ");
+         new_opts.insert(new_opts.begin(), input);
       }
    
       if (!eosio::cdt::environment::exec_subprogram("eosio-ld", new_opts)) {

--- a/tools/cc/eosio-cpp.cpp.in
+++ b/tools/cc/eosio-cpp.cpp.in
@@ -183,26 +183,27 @@ int main(int argc, const char **argv) {
          std::string tmp_file = std::string(res.c_str())+"/"+llvm::sys::path::filename(input).str();
          std::string output;
 
-         generate(opts.comp_options, input, opts.abigen_contract, opts.abigen_resources, opts.abigen);
+         if (!opts.pp_only) {
+            generate(opts.comp_options, input, opts.abigen_contract, opts.abigen_resources, opts.abigen);
 
-         auto src = SmallString<64>(input);
-         llvm::sys::path::remove_filename(src);
-         std::string source_path = src.str().empty() ? "." : src.str();
-         new_opts.insert(new_opts.begin(), "-I" + source_path);
+            auto src = SmallString<64>(input);
+            llvm::sys::path::remove_filename(src);
+            std::string source_path = src.str().empty() ? "." : src.str();
+            new_opts.insert(new_opts.begin(), "-I" + source_path);
 
-         if (llvm::sys::fs::exists(tmp_file)) {
-            input = tmp_file;
+            if (llvm::sys::fs::exists(tmp_file)) {
+               input = tmp_file;
+            }
+            output = tmp_file+".o";
+
+            if (!opts.link) {
+               output = opts.output_fn.empty() ? "a.out" : opts.output_fn;
+            }
+
+            new_opts.insert(new_opts.begin(), {"-o", output});
+            outputs.push_back(output);
          }
-         output = tmp_file+".o";
-
          new_opts.insert(new_opts.begin(), input);
-         
-         if (!opts.link) {
-            output = opts.output_fn.empty() ? "a.out" : opts.output_fn;
-         }
-
-         new_opts.insert(new_opts.begin(), {"-o", output});
-         outputs.push_back(output);
 
          if (llvm::sys::path::extension(input).equals(".c"))
             new_opts.insert(new_opts.begin(), "-xc++");

--- a/tools/cc/eosio-cpp.cpp.in
+++ b/tools/cc/eosio-cpp.cpp.in
@@ -185,7 +185,7 @@ int main(int argc, const char **argv) {
 
          if (!opts.pp_only) {
             auto tool_opts = opts.comp_options;
-            std::set<std::string> non_tool_opts = { "-S", "-emit-llvm" };
+            std::set<std::string> non_tool_opts = { "-S", "-emit-llvm", "-emit-ast" };
             tool_opts.erase(std::remove_if(tool_opts.begin(), tool_opts.end(),
                                            [&](const auto& opt){ return non_tool_opts.count(opt); }),
                             tool_opts.end());

--- a/tools/cc/eosio-cpp.cpp.in
+++ b/tools/cc/eosio-cpp.cpp.in
@@ -184,7 +184,12 @@ int main(int argc, const char **argv) {
          std::string output;
 
          if (!opts.pp_only) {
-            generate(opts.comp_options, input, opts.abigen_contract, opts.abigen_resources, opts.abigen);
+            auto tool_opts = opts.comp_options;
+            std::set<std::string> non_tool_opts = { "-S", "-emit-llvm" };
+            tool_opts.erase(std::remove_if(tool_opts.begin(), tool_opts.end(),
+                                           [&](const auto& opt){ return non_tool_opts.count(opt); }),
+                            tool_opts.end());
+            generate(tool_opts, input, opts.abigen_contract, opts.abigen_resources, opts.abigen);
 
             auto src = SmallString<64>(input);
             llvm::sys::path::remove_filename(src);

--- a/tools/include/compiler_options.hpp.in
+++ b/tools/include/compiler_options.hpp.in
@@ -434,59 +434,64 @@ static void GetLdDefaults(std::vector<std::string>& ldopts) {
       if (fquery_opt || fquery_server_opt || fquery_client_opt) {
          ldopts.emplace_back("--export-table");
          ldopts.emplace_back("-other-model");
-         ldopts.emplace_back("--only-export \"*:table\"");
-         ldopts.emplace_back("--only-export \"*:memory\"");
+         ldopts.insert(ldopts.end(), { "--only-export", "*:table" });
+         ldopts.insert(ldopts.end(), { "--only-export", "*:memory" });
          if (fuse_main_opt) {
-            ldopts.emplace_back("-e main");
-            ldopts.emplace_back("--only-export \"main:function\"");
+            ldopts.insert(ldopts.end(), { "-e", "main" });
+            ldopts.insert(ldopts.end(), { "--only-export", "main:function" } );
          }
          else {
-            ldopts.emplace_back("-e initialize ");
-            ldopts.emplace_back("--only-export \"initialize:function\"");
-         }
-         if (fquery_server_opt)
-            ldopts.emplace_back("-export run_query ");
-         else {
-            ldopts.emplace_back("-export create_query_request ");
-            ldopts.emplace_back("-export decode_query_response ");
-            ldopts.emplace_back("-export describe_query_request ");
-            ldopts.emplace_back("-export describe_query_response ");
+            ldopts.insert(ldopts.end(), { "-e", "initialize " } );
+            ldopts.insert(ldopts.end(), { "--only-export", "initialize:function" } );
          }
          if (fquery_server_opt)
-            ldopts.emplace_back("--only-export \"run_query:function\"");
+            ldopts.insert(ldopts.end(), { "-export", "run_query " });
          else {
-            ldopts.emplace_back("--only-export \"create_query_request:function\"");
-            ldopts.emplace_back("--only-export \"decode_query_response:function\"");
-            ldopts.emplace_back("--only-export \"describe_query_request:function\"");
-            ldopts.emplace_back("--only-export \"describe_query_response:function\"");
+            ldopts.insert(ldopts.end(), { "-export", "create_query_request " });
+            ldopts.insert(ldopts.end(), { "-export", "decode_query_response " });
+            ldopts.insert(ldopts.end(), { "-export", "describe_query_request " });
+            ldopts.insert(ldopts.end(), { "-export", "describe_query_response " });
+         }
+         if (fquery_server_opt)
+            ldopts.insert(ldopts.end(), { "--only-export", "run_query:function" });
+         else {
+            ldopts.insert(ldopts.end(), { "--only-export", "create_query_request:function" });
+            ldopts.insert(ldopts.end(), { "--only-export", "decode_query_response:function" });
+            ldopts.insert(ldopts.end(), { "--only-export", "describe_query_request:function" });
+            ldopts.insert(ldopts.end(), { "--only-export", "describe_query_response:function" });
          }
       } else {
          if (fuse_main_opt)
-            ldopts.emplace_back("-e main");
+            ldopts.insert(ldopts.end(), { "-e", "main" });
          else
-            ldopts.emplace_back("-e apply");
-         ldopts.emplace_back("--only-export \"apply:function\"");
+            ldopts.insert(ldopts.end(), { "-e", "apply" });
+         ldopts.insert(ldopts.end(), { "--only-export", "apply:function" });
       }
-      ldopts.emplace_back("-lc++ -lc -leosio");
+      ldopts.emplace_back("-lc++");
+      ldopts.emplace_back("-lc");
+      ldopts.emplace_back("-leosio");
       if (use_old_malloc_opt)
          ldopts.emplace_back("-leosio_malloc");
       else
          ldopts.emplace_back("-leosio_dsm");
 
-      if (use_rt_opt || fquery_opt || fquery_server_opt || fquery_client_opt)
-         ldopts.emplace_back("-lrt -lsf");
+      if (use_rt_opt || fquery_opt || fquery_server_opt || fquery_client_opt) {
+         ldopts.emplace_back("-lrt");
+         ldopts.emplace_back("-lsf");
+      }
       if (fquery_opt || fquery_server_opt || fquery_client_opt)
          ldopts.emplace_back("-leosio_cmem");
       if (!only_export_opt.empty()) {
-         ldopts.emplace_back("--only-export \""+only_export_opt+"\"");
+         ldopts.emplace_back("--only-export");
+         ldopts.emplace_back(only_export_opt);
       }
 
    } else {
 #ifdef __APPLE__
-      ldopts.emplace_back("-arch x86_64 -macosx_version_min 10.13 -framework Foundation -framework System");
+      ldopts.insert(ldopts.end(), {"-arch", "x86_64", "-macosx_version_min", "10.13", "-framework", "Foundation", "-framework", "System"});
 #endif
       ldopts.emplace_back("-static");
-      ldopts.emplace_back("-lnative_c++ -lnative_c -lnative_eosio -lnative -lnative_rt");
+      ldopts.insert(ldopts.end(), {"-lnative_c++", "-lnative_c", "-lnative_eosio", "-lnative", "-lnative_rt"});
    }
 }
 #endif
@@ -611,7 +616,7 @@ static Options CreateOptions(bool add_defaults=true) {
       copts.emplace_back("-fcolor-diagnostics");
    }
    if (!MF_opt.empty()) {
-      copts.emplace_back("-MF "+MF_opt);
+      copts.insert(copts.end(), { "-MF", MF_opt });
    }
    if (MD_opt) {
       copts.emplace_back("-MD");
@@ -778,10 +783,10 @@ static Options CreateOptions(bool add_defaults=true) {
          llvm::sys::path::replace_extension(fn, "");
          llvm::sys::path::replace_extension(fn, fnative_opt ? "" : ".wasm");
          output_fn = fn.str();
-         ldopts.emplace_back("-o "+output_fn);
+         ldopts.emplace_back("-o"+output_fn);
       } else {
          output_fn = "a.out";
-         ldopts.emplace_back("-o "+output_fn);
+         ldopts.emplace_back("-o"+output_fn);
       }
 #endif
    }

--- a/tools/include/compiler_options.hpp.in
+++ b/tools/include/compiler_options.hpp.in
@@ -354,6 +354,7 @@ struct Options {
    std::vector<std::string> inputs;
    bool link;
    bool abigen;
+   bool pp_only;
    std::string eosio_pp_dir;
    std::string abigen_output;
    std::string abigen_contract;
@@ -511,8 +512,10 @@ static Options CreateOptions(bool add_defaults=true) {
 
 #ifdef ONLY_LD
    bool abigen = false;
+   bool pp_only = false;
 #else
    bool abigen = abigen_opt;
+   bool pp_only = E_opt;
    abigen_output = abigen_output_opt;
    debug = g_opt;
 #endif
@@ -855,8 +858,8 @@ static Options CreateOptions(bool add_defaults=true) {
 #endif
 
 #ifndef ONLY_LD
-   return {output_fn, inputs, link, abigen, pp_dir, abigen_output, abigen_contract, copts, ldopts, agopts, agresources, debug, fnative_opt};
+   return {output_fn, inputs, link, abigen, pp_only, pp_dir, abigen_output, abigen_contract, copts, ldopts, agopts, agresources, debug, fnative_opt};
 #else
-   return {output_fn, {}, link, abigen, pp_dir, abigen_output, abigen_contract, copts, ldopts, agopts, agresources, debug, fnative_opt};
+   return {output_fn, {}, link, abigen, pp_only, pp_dir, abigen_output, abigen_contract, copts, ldopts, agopts, agresources, debug, fnative_opt};
 #endif
 }

--- a/tools/include/compiler_options.hpp.in
+++ b/tools/include/compiler_options.hpp.in
@@ -795,7 +795,8 @@ static Options CreateOptions(bool add_defaults=true) {
 #endif
    }
    else {
-      ldopts.emplace_back("-o "+o_opt);
+      ldopts.emplace_back("-o");
+      ldopts.emplace_back(o_opt);
       output_fn = o_opt;
    }
 

--- a/tools/include/compiler_options.hpp.in
+++ b/tools/include/compiler_options.hpp.in
@@ -600,6 +600,7 @@ static Options CreateOptions(bool add_defaults=true) {
       copts.emplace_back("-D"+define);
    }
    if (emit_ast_opt) {
+      link = false;
       copts.emplace_back("-emit-ast");
    }
    if (emit_llvm_opt) {

--- a/tools/include/eosio/utils.hpp
+++ b/tools/include/eosio/utils.hpp
@@ -134,14 +134,15 @@ struct environment {
      return env_table;
    }
    static bool exec_subprogram(const std::string prog, std::vector<std::string> options, bool root=false) {
-      std::stringstream args;
-      for (auto s : options)
-         args << s << " ";
+      std::vector<llvm::StringRef> args;
+      args.push_back(prog);
+      args.insert(args.end(), options.begin(), options.end());
       std::string find_path = eosio::cdt::whereami::where();
       if (root)
          find_path = "/usr/bin";
-      if ( auto path = llvm::sys::findProgramByName(prog.c_str(), {find_path}) )
-         std::system((*path+" "+args.str()).c_str());
+      if ( auto path = llvm::sys::findProgramByName(prog.c_str(), {find_path}) ) {
+         return llvm::sys::ExecuteAndWait(*path, args, {}, {}, 0, 0, nullptr, nullptr) != -1;
+      }
       else
          return false;
       return true; 

